Services/UICore/classes/class.ilCtrl.php

```diff
@@ -1622,111 +1631,109 @@ private function getClassForCid($a_cid)
 	}
 
 	/**
-	 * Read information of class per cid
-	 * @return 
+	 * Save class respective to $a_cid and store corresponding
+	 * class calls for future reference.
+	 *
 	 * @param object $a_cid		cid
 	 */
 	private function readCidInfo($a_cid)
 	{
-		global $ilDB;
-
 		if (isset($this->info_read_cid[$a_cid]))
 		{
 			return;
 		}
 
 		$cached_ctrl = ilCachedCtrl::getInstance();
-		$rec = $cached_ctrl->lookupCid($a_cid);
+		$cid_info = $cached_ctrl->lookupCid($a_cid);
 
-		if($rec)
+		if($cid_info)
 		{
-			$this->cid_class[$a_cid] = $rec["class"];
-			$this->class_cid[$rec["class"]] = $a_cid;
-
-			$calls = $cached_ctrl->lookupCall($rec["class"]);
+			$this->updateClassCidMap($cid_info['class'], $a_cid);
 
-			foreach($calls as $rec2)
+			foreach($cached_ctrl->lookupCall($cid_info["class"]) as $call)
 			{
-				if (!isset($this->calls[$rec["class"]]) || !is_array($this->calls[$rec["class"]]) || !in_array($rec2["child"], $this->calls[$rec["class"]]))
+				if ($call["child"] != "" && $this->callOfClassNotKnown($cid_info['class'],$call['child']))
 				{
-					if ($rec2["child"] != "")
-					{
-						$this->calls[$rec["class"]][] = $rec2["child"];
-					}
+					$this->calls[$cid_info["class"]][] = $call["child"];
 				}
 			}
-			$this->info_read_class[$rec["class"]] = true;
+			$this->info_read_class[$cid_info["class"]] = true;
 		}
 		
 		$this->info_read_cid[$a_cid] = true;
 	}
 
 	/**
-	 * Read info of node
-	 *  
-	 * @param	object	$a_class	class name
+	 * Save classes respective to the cid's of a node and store corresponding
+	 * class calls for future reference.
+	 *
+	 * @param	string	$a_node
 	 */
 	private function readNodeInfo($a_node)
 	{
-		$n_arr = explode(":", $a_node);
-		foreach ($n_arr as $cid)
+		$node_array = explode(":", $a_node);
```

Comment:
Maybe `$cids` instead of `$node_array`? That is, what these actually are, right?